### PR TITLE
Add native API to hostpolicy that returns the native probe dirs for ASP.NET

### PR DESCRIPTION
**Pull request by steveharter**
Adds a native API to hostfxr.dll that returns the native search paths for a given app. In order to obtain all valid directories (including servicing directories), it must call into the hostpolicy.dll for the current runtime and can't for example simply be implemented just in hostfxr.dll.

Fixes https://github.com/dotnet/core-setup/issues/3458

ASP.NET will call this to address scenarios where they need to pre-load a native dll. ASP.NET will cycle through these returned paths and attempt to load the given native dll manually.

Notes:
- Refactored the fx_muxer.cpp execute-related methods to separate exec, CLI and this new API mode.
- A new hostpolicy.dll entry point was added to address this API mode. It is designed to be used by future API calls in the future (if they just need to return a string).
- The name of the API \ 'command' is not passed to this new hostpolicy.dll entry point. Instead, the name is passed as part of the host_interface_t serialization to allow for future flexibility including adding additional arguments but while keeping the entry point the same.

cc @jkotalik

### CODE REVIEW

**Comment by jkotalik** [comment]
cc/ @muratg @davidfowl

File: src/corehost/cli/fxr/fx_muxer.cpp [hunk 1]
```diff
@@ -1210,17 +1214,7 @@ int fx_muxer_t::parse_args_and_execute(
...
-        vec_argv.reserve(argc - cur_i + 1); // +1 for dotnet
-        vec_argv.push_back(argv[0]);
-        vec_argv.insert(vec_argv.end(), argv + cur_i, argv + argc);
-        new_argv = vec_argv.data();
-        new_argc = vec_argv.size();
-    }
-
-    // Transform dotnet [exec] [--additionalprobingpath path] [--depsfile file] [dll] [args] -> dotnet [dll] [args]
-    return read_config_and_execute(own_dir, app_candidate, opts, new_argc, new_argv, mode);
+    return 1;
```
Comment (eerhardt) [inline src/corehost/cli/fxr/fx_muxer.cpp]: I thought the convention was to return `0` on success.

File: src/corehost/cli/fxr/hostfxr.cpp [hunk 2]
```diff
@@ -178,3 +257,57 @@ SHARED_API int32_t hostfxr_resolve_sdk(
...
+//    buffer_size
+//      The size of the buffer argument in pal::char_t units.
+//
+//    required_buffer_size
+//      If the apis fails because of insufficient buffer size,
+//      required_buffer_size is set to the minimium buffer size
+//      necessary to contain the result.
+//
+// Return value:
+//   1 on success, otherwise failure
```
Comment (eerhardt) [inline src/corehost/cli/fxr/hostfxr.cpp]: I think this should be 0 on success.

File: src/corehost/cli/hostpolicy.cpp [hunk 3]
```diff
@@ -175,7 +187,27 @@ int run(const arguments_t& args)
...
+            trace::error(_X("get-native-search-directories failed to find NATIVE_DLL_SEARCH_DIRECTORIES property"));
+            exit_code = HostApiFailed;
+        }
+        else
+        {
+            assert(out_host_command_result != nullptr);
+            pal::clr_palstring(property_values[1], out_host_command_result);
+        }
+
+        return exit_code;
```
Comment (eerhardt) [inline src/corehost/cli/hostpolicy.cpp]: Should return `0` for success.

File: src/corehost/cli/fxr/fx_muxer.cpp (hunk 1 above)
Comment (steveharter) [reply src/corehost/cli/fxr/fx_muxer.cpp]: Yes. Thanks

File: src/corehost/cli/fxr/hostfxr.cpp (hunk 2 above)
Comment (steveharter) [reply src/corehost/cli/fxr/hostfxr.cpp]: Done

File: src/corehost/cli/hostpolicy.cpp (hunk 3 above)
Comment (steveharter) [reply src/corehost/cli/hostpolicy.cpp]: Done

File: src/corehost/cli/hostpolicy.cpp [hunk 4]
```diff
@@ -273,16 +308,22 @@ int run(const arguments_t& args)
...
 
     return exit_code;
 }
 
 SHARED_API int corehost_load(host_interface_t* init)
 {
     trace::setup();
-    
+
+    // Re-initialize global state in case of re-entry
```
Comment (eerhardt) [inline src/corehost/cli/hostpolicy.cpp]: Do we have a re-entry test case?

Comment (steveharter) [reply src/corehost/cli/hostpolicy.cpp]: Yes, the unit test that was added calls back in a second time because its buffer was too small the first time

**Pull request merged**